lib/pi_piper/sysfs/driver.rb

```diff
@@ -0,0 +1,106 @@
+require 'filewatcher'
+
+module PiPiper
+  module Sysfs
+    class Driver < PiPiper::Driver
+
+      GPIO_HIGH = 1
+      GPIO_LOW  = 0
+
+      def initialize
+        @exported_pins = Set.new
+      end
+
+      def close
+        unexport_all
+        @exported_pins.empty? 
+      end
+      
+  # Support GPIO pins
+      def pin_direction(pin, direction)
+        raise ArgumentError, "direction should be :in or :out" unless [:in, :out].include? direction
+        export(pin)
+        raise RuntimeError, "Pin #{pin} not exported" unless exported?(pin)
+        File.write(direction_file(pin), direction)
+      end
+
+      def pin_read(pin)
+        raise ArgumentError, "Pin #{pin} not exported" unless exported?(pin)
+        File.read(value_file(pin)).to_i
+      end
+
+      def pin_write(pin, value)
+        raise ArgumentError, "value should be GPIO_HIGH or GPIO_LOW" unless [GPIO_LOW, GPIO_HIGH].include? value
+        raise ArgumentError, "Pin #{pin} not exported" unless exported?(pin)
+        File.write(value_file(pin), value)
+      end
+
+      def pin_set_pud(pin, value)
+        raise NotImplementedError, "Pull up/down not available with this driver. keep it on :off" unless value == :off
+      end
+      
+      def pin_set_trigger(pin, trigger)
+        raise ArgumentError, "trigger should be :falling, :rising, :both or :none" unless [:falling, :rising, :both, :none].include? trigger
+        raise ArgumentError, "Pin #{pin} not exported" unless exported?(pin)
+        File.write(edge_file(pin), trigger)
+      end
+
+      def pin_wait_for(pin, trigger)
+        pin_set_trigger(pin, trigger)
+        value = pin_read(pin)
+
```

Comment:
Sure. I thought the original implementation watched the value file and compared whether the value changed or not? And the reason I implemented it this way is because the original implementation actually spins at 100%. I verified this by running the original method manually in an `irb` session. I can double check this again but that was my inspiration. I agree about adding another gem, I wasn't too keen on it but I had difficulty finding some sample code that would block the script until a file write. This gem apparently does it and seems to work well. We should definitely run it through some stuff to verify performance though.

But basically that is what my thinking was. It actually doesn't use File#mtime but rather listens for kernel events. Having said that I did not look at the source code in great detail but the readme mentioned that part. If that is the case then we ultimately hit the kernel anyway so this method may be more robust. I know every OS has a "file system watcher" event handler so hopefully this one is using one in the code.